src/backend/commands/copy.c

```diff
@@ -722,32 +723,84 @@ CopyGetData(CopyState cstate, void *databuf, int datasize)
 	switch (cstate->copy_dest)
 	{
 		case COPY_FILE:
-			bytesread = fread(databuf, 1, datasize, cstate->copy_file);
-			if (feof(cstate->copy_file))
-				cstate->fe_eof = true;
-			if (ferror(cstate->copy_file))
+			if (cstate->is_program)
 			{
-				if (cstate->is_program)
+				struct pollfd pollfd;
+				int			rc;
+				ssize_t		result = 0;
+
+				/* Check if data is available */
+				pollfd.fd = fileno(cstate->copy_file);
+				pollfd.events = POLLIN;
+				pollfd.revents = 0;
+				rc = poll(&pollfd, 1, 0);
+
+				/*
+				 * Continue waiting until we get data, even if interrupted by
+				 * a signal
+				 */
+				while (rc == 0 || (rc < 0 && (errno == EINTR || errno == EAGAIN)))
+				{
+					/*
+					 * It's not available, we have time to check for
+					 * interrupts
+					 */
+					CHECK_FOR_INTERRUPTS();
+
+					/* Wait until data arrives (with 10 second timeout) */
+					pollfd.revents = 0;
+					rc = poll(&pollfd, 1, 10000);
+				}
+
+				/* If data is available */
+				if (rc == 1 && (pollfd.revents & POLLIN))
+				{
+					result = read(pollfd.fd, databuf, datasize);
+					if (result == 0)
+						cstate->fe_eof = true;
+					else if (result > 0)
+						bytesread = result;
+				}
+
+				/* If error occured during poll or read */
+				if (rc < 0 || result < 0)
 				{
-					int olderrno = errno;
+					int			olderrno = errno;
 
 					close_program_pipes(cstate, true);
 
 					/*
-					 * If close_program_pipes() didn't throw an error,
-					 * the program terminated normally, but closed the
-					 * pipe first. Restore errno, and throw an error.
+					 * If close_program_pipes() didn't throw an error, the
+					 * program terminated normally, but closed the pipe first.
+					 * Restore errno, and throw an error.
 					 */
 					errno = olderrno;
 
+					if (rc < 0)
+					{
+						ereport(ERROR,
```

Comment:
I suggest to move error processing closer to the place where error occurred. I.e. right after poll in this case.

Reply:
That would require duplicating the `close_program_pipes` part, is that acceptable? Also, this is supposed to handle errors for both initial `poll` (which is non blocking) and the `poll` inside the loop (which is blocking with a timeout). Should I duplicate this for both of them too?

Reply:
I think code will be cleaner.